### PR TITLE
WASM Plugin for Pixiu

### LINKED ISSUE #432: [ASoC 2022] Pixiu WASM Plugin Support
### Background

This is an basic subject of [[ASoC 2022](https://opensource.alibaba.com/asoc2022)](https://opensource.alibaba.com/asoc2022) and https://github.com/apache/dubbo/issues/10088 .

Pixiu supports the WASM plugin mechanism, dynamically loads and updates WASM plugins, and provides a simple implementation of WASM plugins

## Difficulty

Medium

## Mentor

- @PhilYue 

## Output Requirement

1. Pixiu supports the WASM plugin mechanism
2. Use WASM to implement a simple plug-in that can communicate with the teacher

## 背景

这是一个[[阿里巴巴编程之夏 2022](https://opensource.alibaba.com/asoc2022)](https://opensource.alibaba.com/asoc2022) 的基础课题 https://github.com/apache/dubbo/issues/10088 。

Pixiu 支持动态 WASM 插件机制

Pixiu 支持 WASM 插件机制，能够动态加载和更新 WASM 插件，并且提供一个简易的 WASM 插件实现。

## 难度

中等

## 导师

- @PhilYue 

## 产出要求

1. Pixiu 支持 WASM 插件机制
2. 使用 WASM 实现一个简易的插件，具体插件功能可以和老师沟通

### PR DESCRIPTION
**Pull request by maxingg**
**What this PR does**:
Pixiu supports the WASM plugin mechanism, and provides a simple implementation of WASM plugins.

**Which issue(s) this PR fixes**:

Fixes #432 

**Special notes for your reviewer**:

```release-note
Users can choose to load different wasm files which must be stored in the data folder.
```

### CODE REVIEW

**Review by PhilYue — APPROVED** [review]

**Review by baerwang — APPROVED** [review]

File: pkg/wasm/service.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+	"math"
+	"os"
+	"path/filepath"
+	"sync"
+	"sync/atomic"
+)
+
+import (
+	"github.com/mitchellh/mapstructure"
+
```
Comment (cityiron) [inline pkg/wasm/service.go]: I think not need split here.

File: pkg/wasm/service.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	mutex              sync.Mutex
+	model              model.WasmService
+	instancePool       sync.Pool
+}
+
+type WasmConfig struct {
+	Path string `yaml:"path" json:"path,omitempty"`
+}
+
+type ABIContextWrapper struct {
```
Comment (cityiron) [inline pkg/wasm/service.go:58]: Some comments for this struct? I found some methods are also not commented on.

File: igt/Makefile [hunk 3]
```diff
@@ -49,7 +49,7 @@ else
 	export EXT_NAME ?=
 endif
 
-CGO ?= 0
+CGO ?= 1
```
Comment (mark4z) [inline igt/Makefile:52]: what's this

File: pkg/wasm/service.go (hunk 1 above)
Comment (mark4z) [reply pkg/wasm/service.go]: agree

File: pkg/filter/http/proxywasm/filter.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+func (w *WasmFilter) Decode(ctx *http.HttpContext) filter.FilterStatus {
+	const HEADER = "header"
+
+	// sample: display http header
+	wrapper := w.abiContextWrappers[HEADER]
+	wrapper.Context.Instance.Lock(wrapper.Context)
+	defer wrapper.Context.Instance.Unlock()
+
+	_ = wrapper.Context.GetExports().ProxyOnContextCreate(wrapper.ContextID, wasm.GetServiceRootID(HEADER))
```
Comment (mark4z) [inline pkg/filter/http/proxywasm/filter.go]: maybe error should not ignored

File: igt/Makefile (hunk 3 above)
Comment (maxingg) [reply igt/Makefile:52]: cgo provides a mechanism for golang and C to call each other which is used in wasm-go-sdk, we must active it.

File: pkg/wasm/service.go (hunk 1 above)
Comment (maxingg) [reply pkg/wasm/service.go]: okay

File: pkg/wasm/service.go (hunk 2 above)
Comment (maxingg) [reply pkg/wasm/service.go:58]: Do private methods also need to be commented on?

File: pkg/filter/http/proxywasm/filter.go (hunk 4 above)
Comment (maxingg) [reply pkg/filter/http/proxywasm/filter.go]: okay, I will handle them.

Comment (mark4z) [reply pkg/filter/http/proxywasm/filter.go]: any update?

Comment (maxingg) [reply pkg/filter/http/proxywasm/filter.go]: yes

**Review by mark4z — APPROVED** [review]

File: pkg/filter/http/proxywasm/filter.go [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+	return nil
+}
+
+func (factory *WasmFilterFactory) PrepareFilterChain(ctx *http.HttpContext, chain filter.FilterChain) error {
+	filter := &WasmFilter{
+		abiContextWrappers: make(map[string]*wasm.ABIContextWrapper),
+	}
+	for _, service := range factory.cfg.WasmServices {
+		if abiContext := wasm.CreateABIContextByName(service.Name, ctx); abiContext != nil {
+			filter.abiContextWrappers[service.Name] = abiContext
```
Comment (PhilYue) [inline pkg/filter/http/proxywasm/filter.go:108]: 此处 abiContext 的key `service.Name` 是不是必须要与文件`key.go` 的 `HeaderLevel` 匹配

File: pkg/wasm/service.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+		mutex:              sync.Mutex{},
+		model:              service,
+	}
+	wasmService.rootContextID = atomic.AddInt32(&wasmService.contextIDGenerator, 1)
+
+	wasmService.instancePool = sync.Pool{
+		New: func() interface{} {
+			pwd, _ := os.Getwd()
+			instance := wasmer.NewWasmerInstanceFromFile(filepath.Join(pwd, cfg.Path))
+			proxywasm.RegisterImports(instance)
```
Comment (PhilYue) [inline pkg/wasm/service.go:88]: instance 做非空判断，或者提前判断  path 下 wasm 文件是否存在，给出友好 log

File: pkg/filter/http/proxywasm/filter.go (hunk 5 above)
Comment (maxingg) [reply pkg/filter/http/proxywasm/filter.go:108]: 是的

File: pkg/wasm/service.go (hunk 6 above)
Comment (maxingg) [reply pkg/wasm/service.go:88]: 好

**Pull request merged**